Compile bug: error: more than one conversion function from "half" to a built-in type applies

### Git commit

ef476916bba4b44f44be0c98babc1cb025968e75

### Operating systems

Linux

### GGML backends

CUDA

### Problem description & steps to reproduce

On a fresh vm, compiling with CUDA gives this error. Compiling with CPU works fine.

### First Bad Commit

Unsure which exact commit it started with, but I was not getting errors ~3 days ago

### Compile command

```shell
git clone https://github.com/ggerganov/llama.cpp

pip install -r llama.cpp/requirements.txt

cd llama.cpp

cmake -B build -DGGML_CUDA=ON -DLLAMA_CURL=OFF
cmake --build build --config Release -j 128
```

### Relevant log output

```shell
[ 39%] Building CUDA object ggml/src/ggml-cuda/CMakeFiles/ggml-cuda.dir/template-instances/fattn-vec-f16-instance-hs64-f16-f16.cu.o
[ 39%] Building CUDA object ggml/src/ggml-cuda/CMakeFiles/ggml-cuda.dir/template-instances/fattn-vec-f32-instance-hs256-f16-f16.cu.o
[ 40%] Building CUDA object ggml/src/ggml-cuda/CMakeFiles/ggml-cuda.dir/template-instances/fattn-vec-f32-instance-hs128-f16-f16.cu.o
[ 40%] Building CUDA object ggml/src/ggml-cuda/CMakeFiles/ggml-cuda.dir/template-instances/fattn-vec-f32-instance-hs64-f16-f16.cu.o
[ 41%] Linking CXX shared library ../../bin/libggml-cpu.so
[ 41%] Built target ggml-cpu
/workspace/llama.cpp/ggml/src/ggml-cuda/conv2d.cu(104): error: more than one conversion function from "half" to a built-in type applies:
            function "__half::operator float() const" (declared at line 217 of /usr/local/cuda/include/cuda_fp16.hpp)
            function "__half::operator short() const" (declared at line 235 of /usr/local/cuda/include/cuda_fp16.hpp)
            function "__half::operator unsigned short() const" (declared at line 238 of /usr/local/cuda/include/cuda_fp16.hpp)
            function "__half::operator int() const" (declared at line 241 of /usr/local/cuda/include/cuda_fp16.hpp)
            function "__half::operator unsigned int() const" (declared at line 244 of /usr/local/cuda/include/cuda_fp16.hpp)
            function "__half::operator long long() const" (declared at line 247 of /usr/local/cuda/include/cuda_fp16.hpp)
            function "__half::operator unsigned long long() const" (declared at line 250 of /usr/local/cuda/include/cuda_fp16.hpp)
            function "__half::operator __nv_bool() const" (declared at line 254 of /usr/local/cuda/include/cuda_fp16.hpp)
                  acc += (input_val * kernel_val);
                          ^
          detected during:
            instantiation of "void conv2d_kernel<T,Layout>(const float *, const T *, float *, conv_params) [with T=half, Layout=whcn_layout]" at line 116
            instantiation of "void conv2d_cuda(const float *, const T *, float *, conv_params, cudaStream_t) [with T=half]" at line 120

/workspace/llama.cpp/ggml/src/ggml-cuda/conv2d.cu(104): error: more than one conversion function from "half" to a built-in type applies:
            function "__half::operator float() const" (declared at line 217 of /usr/local/cuda/include/cuda_fp16.hpp)
            function "__half::operator short() const" (declared at line 235 of /usr/local/cuda/include/cuda_fp16.hpp)
            function "__half::operator unsigned short() const" (declared at line 238 of /usr/local/cuda/include/cuda_fp16.hpp)
            function "__half::operator int() const" (declared at line 241 of /usr/local/cuda/include/cuda_fp16.hpp)
            function "__half::operator unsigned int() const" (declared at line 244 of /usr/local/cuda/include/cuda_fp16.hpp)
            function "__half::operator long long() const" (declared at line 247 of /usr/local/cuda/include/cuda_fp16.hpp)
            function "__half::operator unsigned long long() const" (declared at line 250 of /usr/local/cuda/include/cuda_fp16.hpp)
            function "__half::operator __nv_bool() const" (declared at line 254 of /usr/local/cuda/include/cuda_fp16.hpp)
                  acc += (input_val * kernel_val);
                                      ^
          detected during:
            instantiation of "void conv2d_kernel<T,Layout>(const float *, const T *, float *, conv_params) [with T=half, Layout=whcn_layout]" at line 116
            instantiation of "void conv2d_cuda(const float *, const T *, float *, conv_params, cudaStream_t) [with T=half]" at line 120

2 errors detected in the compilation of "/workspace/llama.cpp/ggml/src/ggml-cuda/conv2d.cu".
gmake[2]: *** [ggml/src/ggml-cuda/CMakeFiles/ggml-cuda.dir/build.make:230: ggml/src/ggml-cuda/CMakeFiles/ggml-cuda.dir/conv2d.cu.o] Error 1
gmake[2]: *** Waiting for unfinished jobs....
gmake[1]: *** [CMakeFiles/Makefile2:1817: ggml/src/ggml-cuda/CMakeFiles/ggml-cuda.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2
```

--- a/ggml/src/ggml-cuda/conv2d.cu
+++ b/ggml/src/ggml-cuda/conv2d.cu
@@ -82,7 +82,7 @@ static __global__ void conv2d_kernel(const float * __restrict__ input,
     int64_t n, c_out, out_y, out_x;
     Layout::unpack_indices(global_idx, P, n, c_out, out_y, out_x);
 
-    T acc = 0;
+    float acc = 0.0f;
 
     for (int64_t c_in = 0; c_in < P.IC; ++c_in) {
         kernel_bounds bounds = calculate_kernel_bounds(out_x, out_y, P);
@@ -93,21 +93,15 @@ static __global__ void conv2d_kernel(const float * __restrict__ input,
             for (int64_t kx = bounds.x_min; kx < bounds.x_max; ++kx) {
                 const int64_t in_x = calculate_input_coord(out_x, kx, P.ST_X, P.DL_X, P.PD_X);
 
-                T input_val;
-                if (std::is_same<T, half>::value) {
-                    input_val = __float2half(input[Layout::input_index(n, c_in, in_y, in_x, P)]);
-                } else {
-                    input_val = input[Layout::input_index(n, c_in, in_y, in_x, P)];
-                }
-
-                T kernel_val = kernel[Layout::kernel_index(c_out, c_in, ky, kx, P)];
+                const float input_val = input[Layout::input_index(n, c_in, in_y, in_x, P)];
+                const float kernel_val = kernel[Layout::kernel_index(c_out, c_in, ky, kx, P)];
                 acc += (input_val * kernel_val);
             }
         }
     }
 
     // [N, OC, OH, OW]
-    output[Layout::output_index(n, c_out, out_y, out_x, P)] = (float) acc;
+    output[Layout::output_index(n, c_out, out_y, out_x, P)] = acc;
 }
 
 template <typename T>

